File: libptiff/src/io/backend/tiff/ptiff_metadata.cpp
#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <span>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

#include <ptiff/core/error.hpp>
#include <ptiff/core/result.hpp>
#include <ptiff/io/backend/tiff/ptiff_metadata.hpp>
#include <ptiff/io/storage_model.hpp>

namespace ptiff::io::backend::tiff {

namespace {

void writeU16(std::vector<std::byte>& out, std::uint16_t v) {
    out.push_back(static_cast<std::byte>(v & 0xFF));
    out.push_back(static_cast<std::byte>((v >> 8) & 0xFF));
}

void writeU32(std::vector<std::byte>& out, std::uint32_t v) {
    for (unsigned i = 0; i < 4; ++i) {
        out.push_back(static_cast<std::byte>((v >> (8 * i)) & 0xFF));
    }
}

void appendAscii(std::vector<std::byte>& out, std::string_view s) {
    for (char c : s) {
        out.push_back(static_cast<std::byte>(static_cast<unsigned char>(c)));
    }
}

std::uint16_t readU16(std::span<const std::byte> s) {
    std::uint16_t v = 0;
    for (unsigned i = 0; i < 2; ++i) {
        v |= static_cast<std::uint16_t>(static_cast<uint8_t>(s[i])) << (8 * i);
    }
    return v;
}

std::uint32_t readU32(std::span<const std::byte> s) {
    std::uint32_t v = 0;
    for (unsigned i = 0; i < 4; ++i) {
        v |= static_cast<std::uint32_t>(static_cast<uint8_t>(s[i])) << (8 * i);
    }
    return v;
}

std::string asciiOf(std::span<const std::byte> s) {
    std::string out;
    out.reserve(s.size());
    for (const auto b : s) {
        out.push_back(static_cast<char>(static_cast<unsigned char>(b)));
    }
    return out;
}

bool hasMagic(std::span<const std::byte> data) {
    if (data.size() < kPtiffMagic.size()) {
        return false;
    }
    for (std::size_t i = 0; i < kPtiffMagic.size(); ++i) {
        if (data[i] != kPtiffMagic[i]) {
            return false;
        }
    }
    return true;
}

} // namespace

Result<std::vector<std::byte>> encodeMetadataPayload(const std::vector<MetadataRecord>& records) {
    // Canonical order: sort by key (byte-wise). Value ties keep a stable relative order via
    // std::stable_sort so the encoding is deterministic even for duplicate keys.
    auto sorted = records;
    std::stable_sort(
        sorted.begin(), sorted.end(), [](const MetadataRecord& a, const MetadataRecord& b) {
            return a.key < b.key;
        });

    std::vector<std::byte> out;
    out.reserve(7 + 8 * sorted.size());
    for (auto b : kPtiffMagic) {
        out.push_back(b);
    }
    writeU16(out, kPtiffMetadataVersion);
    if (sorted.size() > std::numeric_limits<std::uint32_t>::max()) {
        return std::unexpected(
            Error{ErrorCode::InvalidArgument, "encodeMetadataPayload: too many records"});
    }
    writeU32(out, static_cast<std::uint32_t>(sorted.size()));
    for (const auto& rec : sorted) {
        if (rec.key.size() > std::numeric_limits<std::uint16_t>::max()) {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument,
                      "encodeMetadataPayload: metadata key longer than 65535 bytes"});
        }
        if (rec.value.size() > std::numeric_limits<std::uint32_t>::max()) {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument,
                      "encodeMetadataPayload: metadata value longer than 4 GiB"});
        }
        writeU16(out, static_cast<std::uint16_t>(rec.key.size()));
        appendAscii(out, rec.key);
        writeU32(out, static_cast<std::uint32_t>(rec.value.size()));
        appendAscii(out, rec.value);
    }
    return out;
}

Result<std::vector<MetadataRecord>>
decodeMetadataPayload(const std::vector<std::uint64_t>& byteValues) {
    // Converge the reader's `vector<uint64_t>` byte values into a `span<const byte>`.
    std::vector<std::byte> bytes;
    bytes.reserve(byteValues.size());
    for (const auto v : byteValues) {
        if (v > 0xFF) {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument,
                      "decodeMetadataPayload: non-byte value in PTIFF extension tag"});
        }
        bytes.push_back(static_cast<std::byte>(v & 0xFF));
    }
    const std::span<const std::byte> data{bytes};

    if (!hasMagic(data)) {
        return std::unexpected(
            Error{ErrorCode::InvalidArgument,
                  "decodeMetadataPayload: missing PTIFF magic in extension tag payload"});
    }
    std::size_t cursor = kPtiffMagic.size();
    if (data.size() < cursor + 2) {
        return std::unexpected(
            Error{ErrorCode::InvalidArgument,
                  "decodeMetadataPayload: truncated PTIFF extension payload header"});
    }
    const auto version = readU16(data.subspan(cursor, 2));
    cursor += 2;
    if (version > kPtiffMetadataVersion) {
        // A payload from a future writer: do not mis-parse. Return "unsupported version" so the
        // caller can choose to treat the tag as absent (it must not fail the whole file).
        return std::unexpected(
            Error{ErrorCode::InvalidArgument,
                  "decodeMetadataPayload: unsupported PTIFF extension payload version"});
    }
    if (data.size() < cursor + 4) {
        return std::unexpected(
            Error{ErrorCode::InvalidArgument,
                  "decodeMetadataPayload: truncated PTIFF extension payload header"});
    }
    const auto recCount = readU32(data.subspan(cursor, 4));
    cursor += 4;

    std::vector<MetadataRecord> records;
    records.reserve(recCount);
    for (std::uint32_t i = 0; i < recCount; ++i) {
        if (data.size() < cursor + 2) {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument,
                      "decodeMetadataPayload: truncated PTIFF extension record key length"});
        }
        const auto keyLen = readU16(data.subspan(cursor, 2));
        cursor += 2;
        if (data.size() < cursor + keyLen) {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument,
                      "decodeMetadataPayload: truncated PTIFF extension record key"});
        }
        const std::string key = asciiOf(data.subspan(cursor, keyLen));
        cursor += keyLen;

        if (data.size() < cursor + 4) {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument,
                      "decodeMetadataPayload: truncated PTIFF extension record value length"});
        }
        const auto valLen = readU32(data.subspan(cursor, 4));
        cursor += 4;
        if (data.size() < cursor + valLen) {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument,
                      "decodeMetadataPayload: truncated PTIFF extension record value"});
        }
        const std::string value = asciiOf(data.subspan(cursor, valLen));
        cursor += valLen;

        records.push_back(MetadataRecord{std::move(key), std::move(value)});
    }
    return records;
}

std::vector<MetadataRecord> recordsFromStorageModel(const StorageModel& node,
                                                    std::string_view prefix) {
    // StorageModel fields are already stored in a std::map (ascending key order). Collect any
    // field whose key starts with `ptiff.<prefix>.` and strip that prefix. The read path writes
    // with the same prefix so the roundtrip is symmetric.
    const std::string pfx = "ptiff." + std::string(prefix) + ".";
    std::vector<MetadataRecord> records;
    node.for_each_field([&](std::string_view key, std::string_view value) {
        if (key.size() > pfx.size() && key.substr(0, pfx.size()) == pfx) {
            records.push_back(
                MetadataRecord{std::string{key.substr(pfx.size())}, std::string{value}});
        }
    });
    // recordsFromStorageModel preserves the stable ascending key order from the StorageModel map.
    return records;
}

} // namespace ptiff::io::backend::tiff

File: crates/ptiff-core/src/io/backend/tiff/ptiff_metadata.rs
//! PTIFF extension metadata (RFC-7002): the versioned binary payload carried
//! by the five private TIFF tags 65001-65005.
//!
//! Mirrors `ptiff::io::backend::tiff::ptiff_metadata.{hpp,cpp}`.

use crate::io::StorageModel;
use crate::{Error, Result};

/// PTIFF extension payload format version. Readers encountering a version they
/// do not know MUST NOT fail the whole file -- they skip the tag (treating the
/// domain as absent), exactly as a non-PTIFF reader already ignores the private
/// tags.
pub const K_PTIFF_METADATA_VERSION: u16 = 1;

/// ASCII magic bytes `PTIFF` written at the start of every private-tag payload
/// so the decoder can distinguish a PTIFF extension payload from arbitrary
/// bytes left in a private tag by other writers.
pub const K_PTIFF_MAGIC: [u8; 5] = *b"PTIFF";

/// One human-readable (key, value) record.
///
/// Keys select the well-known per-domain field names; unknown keys are
/// preserved verbatim so a future writer's additional fields survive a
/// read-modify-write round trip.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MetadataRecord {
    /// The record key (domain field name).
    pub key: String,
    /// The record value.
    pub value: String,
}

impl MetadataRecord {
    /// Builds a new record.
    #[must_use]
    pub const fn new(key: String, value: String) -> Self {
        Self { key, value }
    }
}

fn write_u16(out: &mut Vec<u8>, v: u16) {
    out.push((v & 0xFF) as u8);
    out.push(((v >> 8) & 0xFF) as u8);
}

fn write_u32(out: &mut Vec<u8>, v: u32) {
    out.push((v & 0xFF) as u8);
    out.push(((v >> 8) & 0xFF) as u8);
    out.push(((v >> 16) & 0xFF) as u8);
    out.push(((v >> 24) & 0xFF) as u8);
}

fn append_ascii(out: &mut Vec<u8>, s: &str) {
    out.extend_from_slice(s.as_bytes());
}

fn read_u16(s: &[u8]) -> u16 {
    let mut v: u16 = 0;
    for (i, &b) in s.iter().take(2).enumerate() {
        v |= u16::from(b) << (8 * i);
    }
    v
}

fn read_u32(s: &[u8]) -> u32 {
    let mut v: u32 = 0;
    for (i, &b) in s.iter().take(4).enumerate() {
        v |= u32::from(b) << (8 * i);
    }
    v
}

fn has_magic(data: &[u8]) -> bool {
    data.len() >= K_PTIFF_MAGIC.len() && data[..K_PTIFF_MAGIC.len()] == K_PTIFF_MAGIC
}

/// Encodes a set of (key, value) records into a byte payload for the given tag.
///
/// Records are sorted by key (ascending, byte-wise) so the encoding is
/// canonical/deterministic: the same logical metadata always yields the same
/// bytes, which is what the golden digest depends on.
///
/// # Errors
///
/// Returns [`crate::ErrorCode::InvalidArgument`] on key/value length overflow
/// (> 65535 / > 4 GiB respectively) or an excessive record count.
pub fn encode_metadata_payload(records: &[MetadataRecord]) -> Result<Vec<u8>> {
    // Canonical order: sort by key (byte-wise), keeping ties stable so the
    // encoding is deterministic even for duplicate keys.
    let mut sorted = records.to_vec();
    sorted.sort_by(|a, b| a.key.cmp(&b.key));

    let mut out = Vec::with_capacity(7 + 8 * sorted.len());
    out.extend_from_slice(&K_PTIFF_MAGIC);
    write_u16(&mut out, K_PTIFF_METADATA_VERSION);
    if sorted.len() > u32::MAX as usize {
        return Err(Error::invalid_argument(
            "encodeMetadataPayload: too many records",
        ));
    }
    write_u32(&mut out, sorted.len() as u32);
    for rec in &sorted {
        if rec.key.len() > u16::MAX as usize {
            return Err(Error::invalid_argument(
                "encodeMetadataPayload: metadata key longer than 65535 bytes",
            ));
        }
        if rec.value.len() > u32::MAX as usize {
            return Err(Error::invalid_argument(
                "encodeMetadataPayload: metadata value longer than 4 GiB",
            ));
        }
        write_u16(&mut out, rec.key.len() as u16);
        append_ascii(&mut out, &rec.key);
        write_u32(&mut out, rec.value.len() as u32);
        append_ascii(&mut out, &rec.value);
    }
    Ok(out)
}

/// Decodes a byte payload back into its (key, value) records.
///
/// # Errors
///
/// Returns [`crate::ErrorCode::InvalidArgument`] if the payload is empty, lacks
/// the PTIFF magic, has a version > [`K_PTIFF_METADATA_VERSION`], is
/// malformed/truncated, or contains a non-byte value. A payload validated here
/// is guaranteed to round-trip.
pub fn decode_metadata_payload(byte_values: &[u64]) -> Result<Vec<MetadataRecord>> {
    // Converge the reader's `u64` byte values into a plain byte buffer.
    let mut bytes = Vec::with_capacity(byte_values.len());
    for &v in byte_values {
        if v > 0xFF {
            return Err(Error::invalid_argument(
                "decodeMetadataPayload: non-byte value in PTIFF extension tag",
            ));
        }
        bytes.push(v as u8);
    }
    let data = &bytes;

    if !has_magic(data) {
        return Err(Error::invalid_argument(
            "decodeMetadataPayload: missing PTIFF magic in extension tag payload",
        ));
    }
    let mut cursor = K_PTIFF_MAGIC.len();
    if data.len() < cursor + 2 {
        return Err(Error::invalid_argument(
            "decodeMetadataPayload: truncated PTIFF extension payload header",
        ));
    }
    let version = read_u16(&data[cursor..cursor + 2]);
    cursor += 2;
    if version > K_PTIFF_METADATA_VERSION {
        // A payload from a future writer: do not mis-parse.
        return Err(Error::invalid_argument(
            "decodeMetadataPayload: unsupported PTIFF extension payload version",
        ));
    }
    if data.len() < cursor + 4 {
        return Err(Error::invalid_argument(
            "decodeMetadataPayload: truncated PTIFF extension payload header",
        ));
    }
    let rec_count = read_u32(&data[cursor..cursor + 4]);
    cursor += 4;

    let mut records = Vec::with_capacity(rec_count as usize);
    for _ in 0..rec_count {
        if data.len() < cursor + 2 {
            return Err(Error::invalid_argument(
                "decodeMetadataPayload: truncated PTIFF extension record key length",
            ));
        }
        let key_len = read_u16(&data[cursor..cursor + 2]) as usize;
        cursor += 2;
        if data.len() < cursor + key_len {
            return Err(Error::invalid_argument(
                "decodeMetadataPayload: truncated PTIFF extension record key",
            ));
        }
        let key = String::from_utf8_lossy(&data[cursor..cursor + key_len]).into_owned();
        cursor += key_len;

        if data.len() < cursor + 4 {
            return Err(Error::invalid_argument(
                "decodeMetadataPayload: truncated PTIFF extension record value length",
            ));
        }
        let val_len = read_u32(&data[cursor..cursor + 4]) as usize;
        cursor += 4;
        if data.len() < cursor + val_len {
            return Err(Error::invalid_argument(
                "decodeMetadataPayload: truncated PTIFF extension record value",
            ));
        }
        let value = String::from_utf8_lossy(&data[cursor..cursor + val_len]).into_owned();
        cursor += val_len;

        records.push(MetadataRecord::new(key, value));
    }
    Ok(records)
}

/// Converts the (key, value) archive of a PTIFF extension tag into a
/// [`StorageModel`] node's fields, or reverse: reads every `ptiff.<domain>.*`
/// field of `node`, strips the prefix, and returns the ordered record list to
/// hand to [`encode_metadata_payload`]. `prefix` is e.g. `"spice"` (so the
/// StorageModel holds `ptiff.spice.frame`).
///
/// [`StorageModel`]: crate::io::StorageModel
#[must_use]
pub fn records_from_storage_model(node: &StorageModel, prefix: &str) -> Vec<MetadataRecord> {
    // StorageModel fields are already stored in a BTreeMap (ascending key
    // order). Collect any field whose key starts with `ptiff.<prefix>.` and
    // strip that prefix; the read path writes with the same prefix so the
    // roundtrip is symmetric.
    let pfx = format!("ptiff.{prefix}.");
    let mut records = Vec::new();
    node.for_each_field(|key, value| {
        if key.len() > pfx.len() && key.starts_with(&pfx) {
            records.push(MetadataRecord::new(
                key[pfx.len()..].to_string(),
                value.to_string(),
            ));
        }
    });
    // Preserves the stable ascending key order from the StorageModel map.
    records
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::io::StorageModel;

    fn rec(key: &str, value: &str) -> MetadataRecord {
        MetadataRecord::new(key.to_string(), value.to_string())
    }

    #[test]
    fn encode_is_canonical_and_sorted() {
        let payload =
            encode_metadata_payload(&[rec("b", "2"), rec("a", "1"), rec("c", "3")]).unwrap();
        // Header: magic(5) + version(2) + count(4) = 11.
        // 11 header bytes + 3 records * 8 bytes each (klen2+k1+vlen4+v1) = 35.
        assert_eq!(payload.len(), 35);
        assert_eq!(&payload[0..5], b"PTIFF");
        assert_eq!(read_u16(&payload[5..7]), 1);
        assert_eq!(read_u32(&payload[7..11]), 3);
        // First record: klen (u16 LE) = 1 at [11..13], key 'a' at [13].
        assert_eq!(read_u16(&payload[11..13]), 1); // klen
        assert_eq!(payload[13], b'a');
    }

    #[test]
    fn encode_rejects_oversized_key() {
        let long_key = "k".repeat(65536);
        let err = encode_metadata_payload(&[rec(&long_key, "v")]).unwrap_err();
        assert_eq!(err.code(), crate::ErrorCode::InvalidArgument);
    }

    #[test]
    fn decode_round_trip() {
        let records = vec![rec("frame", "MOUNT_SPACECRAFT"), rec("target", "MARS")];
        let payload = encode_metadata_payload(&records).unwrap();
        let byte_values: Vec<u64> = payload.iter().map(|&b| u64::from(b)).collect();
        let decoded = decode_metadata_payload(&byte_values).unwrap();
        // Sorted ascending by key -> frame, target.
        assert_eq!(
            decoded,
            vec![rec("frame", "MOUNT_SPACECRAFT"), rec("target", "MARS")]
        );
    }

    #[test]
    fn decode_rejects_non_byte_value() {
        let err = decode_metadata_payload(&[0, 0x100]).unwrap_err();
        assert_eq!(err.code(), crate::ErrorCode::InvalidArgument);
    }

    #[test]
    fn decode_rejects_missing_magic() {
        let err = decode_metadata_payload(&[0, 1, 2, 3, 4, 5]).unwrap_err();
        assert_eq!(err.code(), crate::ErrorCode::InvalidArgument);
    }

    #[test]
    fn decode_rejects_future_version() {
        // Valid magic, version = 99.
        let mut payload = K_PTIFF_MAGIC.to_vec();
        write_u16(&mut payload, 99);
        write_u32(&mut payload, 0);
        let byte_values: Vec<u64> = payload.iter().map(|&b| u64::from(b)).collect();
        let err = decode_metadata_payload(&byte_values).unwrap_err();
        assert_eq!(err.code(), crate::ErrorCode::InvalidArgument);
    }

    #[test]
    fn decode_rejects_truncated() {
        let records = vec![rec("a", "longvalue")];
        let payload = encode_metadata_payload(&records).unwrap();
        let byte_values: Vec<u64> = payload.iter().map(|&b| u64::from(b)).collect();
        // Truncate the trailing value.
        let err = decode_metadata_payload(&byte_values[..byte_values.len() - 3]).unwrap_err();
        assert_eq!(err.code(), crate::ErrorCode::InvalidArgument);
    }

    #[test]
    fn records_from_storage_model_strips_prefix() {
        let mut m = StorageModel::new();
        m.set_field("ptiff.spice.frame", "MOUNT");
        m.set_field("ptiff.spice.target", "MARS");
        m.set_field("imageWidth", "64"); // unrelated
        let records = records_from_storage_model(&m, "spice");
        assert_eq!(records, vec![rec("frame", "MOUNT"), rec("target", "MARS")]);
    }

    #[test]
    fn records_from_storage_model_ignores_other_domains() {
        let mut m = StorageModel::new();
        m.set_field("ptiff.camera.k", "fx");
        let records = records_from_storage_model(&m, "spice");
        assert!(records.is_empty());
    }
}
